packages/luals-annotations/src/annotationBuilder.ts

```diff
@@ -0,0 +1,104 @@
+import { Docs } from "./json-docs";
+import luaLSType from "./luaLSType";
+
+export default class AnnotationBuilder {
+	generatedAnnotations: string[] = [];
+
+	currentClass: string = "";
+	staticMethod: boolean = false;
+
+	constructor() {}
+
+	add(...lines: string[]) {
+		this.generatedAnnotations.push(...lines);
+	}
+
+	addComment(comment: string) {
+		this.add(`--- ${comment}`);
+	}
+
+	addDescription(description: string | string[]) {
+		if (typeof description === "string") {
+			this.addComment(description);
+		} else {
+			description.forEach((line) => this.addComment(line));
+		}
+	}
+
+	addFunction(func: Docs.Function) {
+		const params = func.parameters ?? [];
+		const returns = func.returns ?? [];
+
+		const paramList = params.map(
+			(param) => `---@param ${param.name}${(param.optional && "?") || ""} ${luaLSType(param)}`
+		);
+		const returnList = returns.map((ret) => `---@return ${luaLSType(ret)} ${ret.name ?? ""}`);
+
+		const functionCharacter = this.staticMethod ? "." : ":";
+		const functionPrefix = this.currentClass ? `${this.currentClass}${functionCharacter}` : "";
+
+		const renderFunction = (signature: string) => {
+			this.addDescription(func.description);
+			this.add(...paramList, ...returnList, signature);
+		};
```

Comment:
We should generally avoid defining functions inside other functions, except when writing a higher order function.